File: src/lib/contracts/arcane/state-paths.mjs
import { join } from 'node:path';
import { digestValue } from './canonical.mjs';

export const stateRoot = (workspace) => join(workspace, '.audit', 'arcane');
export const statePaths = (workspace) => {
  const root = stateRoot(workspace);
  return { root, receipts: join(root, 'receipts'), replay: join(root, 'replay'), capabilityGrants: join(root, 'capabilities', 'grants'), capabilityTransitions: join(root, 'capabilities', 'transitions'), contractSeals: join(root, 'contract-seals'), authorityBindings: join(root, 'authority-bindings'), sessionBindings: join(root, 'session-bindings'), preEffectCorrelations: join(root, 'pre-effect-correlations') };
};
export const keyHex = (domain, values) => digestValue({ domain, values }).slice(7);
export const stateFile = (dir, domain, values) => join(dir, `${keyHex(domain, values)}.json`);

File: engine/crates/legion-arcane/src/state_paths.rs
use legion_contracts::canonical_digest;
use serde_json::json;
use std::path::{Path, PathBuf};

pub fn state_root(workspace: &Path) -> PathBuf {
    workspace.join(".audit").join("arcane")
}

pub fn key_hex(domain: &str, values: &[String]) -> Result<String, legion_contracts::canonical::CanonicalError> {
    let digest = canonical_digest(&json!({ "domain": domain, "values": values }))?;
    Ok(digest
        .strip_prefix("sha256:")
        .unwrap_or(digest.as_str())
        .to_owned())
}

pub fn state_file(
    dir: &Path,
    domain: &str,
    values: &[String],
) -> Result<PathBuf, legion_contracts::canonical::CanonicalError> {
    Ok(dir.join(format!("{}.json", key_hex(domain, values)?)))
}
